### PR TITLE
feat(metrics): Add `weaviate_build_info` prometheus metrics.

### PR DESCRIPTION
**Pull request by kavirajk**
### What's being changed:

Rationale:
1. Good to group panics/bugs if we tag each binary with build info
2. Having it as prometheus metrics let us to group Grafana alerts specific to weaviate versions.

This is what it looks like
```
weaviate_build_info{branch="main",goarch="arm64",goos="darwin",goversion="go1.23.1",revision="23f2a157c",tags="unknown",version="v1.27.0-alpha"} 1
```

P.S: We can also remove in-built `tags` label at the scraping phase if needed.

Also moved some existing tags from `config` package to `build` package, it made sense. Updated some log entries to use these info from `build` package instead.

Issue: #5873 

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [ ] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

File: tools/generate-release-artifacts.sh [hunk 1]
```diff
@@ -36,19 +43,22 @@ function step_complete() {
 }
 
 function build_binary_arm64() {
-  build_binary "arm64"
+  build_binary "linux" "arm64"
+  build_binary "darwin" "arm64"
```
Comment (kavirajk) [inline tools/generate-release-artifacts.sh:47]: Added support to build artifacts also for `darwin` as well, because I wanted to test this changes on my laptop. Also made sense to have these artifacts along with `linux`.

If anyone have any historic context or conflict on these. Happy to know :)

File: usecases/config/config_handler.go [hunk 2]
```diff
@@ -40,13 +40,6 @@ import (
 // spec only needs to be parsed once.
 var ServerVersion string
```
Comment (kavirajk) [inline usecases/config/config_handler.go:41]: Now I don't know why we need this?

How is it different than `Version`?

File: adapters/handlers/rest/configure_api.go [hunk 3]
```diff
@@ -588,8 +595,8 @@ func configureAPI(api *operations.WeaviateAPI) http.Handler {
 	appState := MakeAppState(ctx, connectorOptionGroup)
 
 	appState.Logger.WithFields(logrus.Fields{
-		"server_version":   config.ServerVersion,
-		"docker_image_tag": config.ImageTag,
+		"server_version": config.ServerVersion,
+		"version":        build.Version,
```
Comment (kavirajk) [inline adapters/handlers/rest/configure_api.go:599]: I replaced `docker_image_tag` -> just simple `version` because

1. It works for every case, even we are running it as container or normal local process. Exposing "docker_*` seems irrelevant.
2. Plus, [what we set for `docker_image_tag` even before it's indeed the version.](https://github.com/weaviate/weaviate/blob/main/tools/generate-release-artifacts.sh#L51)

File: usecases/build/build.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+// These package level variables are set during build time via ldflags -X
+// Also these information are passed to prometheus version package so that it's available in nice `weaviate_info` metric
+var (
+	Version   string
+	Revision  string
+	Branch    string
+	BuildUser string
+	BuildDate string
+	GoVersion string
```
Comment (antas-marcin) [inline usecases/build/build.go:28]: I think it would be great to add comments on each setting what it means

File: tools/generate-release-artifacts.sh [hunk 5]
```diff
@@ -36,19 +43,22 @@ function step_complete() {
...
 }
 
 function build_binary_amd64() {
-  build_binary "amd64"
+  build_binary "linux" "amd64"
+  build_binary "darwin" "amd64"
```
Comment (antas-marcin) [inline tools/generate-release-artifacts.sh:52]: we don't use this script to generate darwin binaries, this script is only used when the CI pipeline that generates release binaries (linux/windows) fails. We have a separate script that generates signed mac binaries, so I would remove this line

File: tools/generate-release-artifacts.sh (hunk 1 above)
Comment (antas-marcin) [reply tools/generate-release-artifacts.sh:47]: added comment in the next file, I would remove this line

File: tools/dev/run_dev_server.sh [hunk 6]
```diff
@@ -20,11 +20,18 @@ export CLUSTER_HOSTNAME=${CLUSTER_HOSTNAME:-"weaviate-0"}
 export GPT4ALL_INFERENCE_API="http://localhost:8010"
 export DISABLE_TELEMETRY=true # disable telemetry for local development
 
+# inject build info into binaries.
+GIT_REVISION=$(git rev-parse --short HEAD)
+GIT_BRANCH=$(git rev-parse --abbrev-ref HEAD)
+
+VERSION="v$(jq -r '.info.version' openapi-specs/schema.json)"
```
Comment (antas-marcin) [inline tools/dev/run_dev_server.sh:27]: in configure_api.go we have a method called `parseVersionFromSwaggerSpec()` that parses the current version, this line is not needed

File: usecases/build/build.go [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+	"github.com/prometheus/common/version"
+)
+
+// These package level variables are set during build time via ldflags -X
+// Also these information are passed to prometheus version package so that it's available in `weaviate_build_info` metric
+var (
+	// Version is Weviate's version. e.g: 1.26.4.
+	// We usually read it from `openapi-specs/schema.json` in this repository.
+	Version string
```
Comment (antas-marcin) [inline usecases/build/build.go:25]: this part will sometimes be the version and sometimes a preview image tag... I would separate those to, as it was to `version` and `imageTag`

File: usecases/build/build.go [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+	// Revision represents the GIT commit of the build.
+	Revision string
```
Comment (antas-marcin) [inline usecases/build/build.go:28]: I'm thinking of be

File: .github/workflows/pull_requests.yaml [hunk 9]
```diff
@@ -327,7 +327,7 @@ jobs:
           install-only: true
       - name: goreleaser
         run: |
-          GIT_HASH=$(git rev-parse --short HEAD) goreleaser build  --clean --snapshot
+          WEAVIATE_GIT_REVISION=$(git rev-parse --short HEAD) WEAVIATE_GIT_BRANCH=$(git rev-parse --abbrev-ref HEAD) WEAVIATE_VERSION=$(git describe --tag) WEAVIATE_BUILD_USER=$(whoami)@$(hostname) WEAVIATE_BUILD_DATE=$(date -u +"%Y-%m-%dT%H:%M:%SZ") goreleaser build  --clean --snapshot
```
[the remark this answers is not in the document]
Comment (kavirajk) [reply .github/workflows/pull_requests.yaml:330]: The rationale for `WEAVIATE_ ` prefix is, [when we export in the shell, we are sharing it with other process](https://github.com/weaviate/weaviate/blob/kavirajk%2Fweaviate-build-info/tools/dev/goreleaser_and_sign.sh#L9-L15). These variables in `pull_requests.yaml`. will be supplied to `.goreleaser.yaml` [here](https://github.com/weaviate/weaviate/blob/kavirajk%2Fweaviate-build-info/.goreleaser.yaml#L16-L23) I believe?

So I think better to prefix, that is unique to the process, coz we don't control what other process that may share that shell and just exporting ENVs like VERSION or BRANCH may mean different things to different process( same way all ENV related to Golang, prefix with GO_ prefix for example). 
If it's just local variable just inside the script or go file, we don't need prefix, shorter local variable is better (that's why no need for `WEAVIATE_` prefix on other places that are local to script. 

wdyt?

File: tools/dev/run_dev_server.sh (hunk 6 above)
Comment (kavirajk) [reply tools/dev/run_dev_server.sh:27]: I took inspiration from [here. ](https://github.com/weaviate/weaviate/blob/main/ci/push_docker.sh#L23). May be not needed for `run_dev_server`. happy to remove if you can confirm :)

File: usecases/build/build.go (hunk 8 above)
Comment (kavirajk) [reply usecases/build/build.go:28]: I don't have too much opinion here. Just inspired from prometheus version package.
https://github.com/prometheus/common/blob/main/version/info.go#L27-L37

Sounds good to change to `GitRevsion`, `GitBranch`.

File: tools/generate-release-artifacts.sh (hunk 5 above)
Comment (kavirajk) [reply tools/generate-release-artifacts.sh]: Done.

File: usecases/build/build.go (hunk 4 above)
Comment (kavirajk) [reply usecases/build/build.go]: Done.

File: ci/push_docker.sh [hunk 10]
```diff
@@ -2,62 +2,41 @@
...
-  if [ "$git_branch" == "main" ] || [ "$git_branch" == "stable/v*" ]; then
-    prefix="$(echo $git_branch | sed 's/\//-/g')"
-  fi
 
+  git_revision=$(echo "$GITHUB_SHA" | cut -c1-7)
+  git_branch=$(git branch) # we clone only specific branch, so $(git branch) will have single value always.
+  build_user="ci"
+  build_date=$(date -u +"%Y-%m-%dT%H:%M:%SZ")
 
   weaviate_version="$(jq -r '.info.version' < openapi-specs/schema.json)"
```
Comment (antas-marcin) [inline ci/push_docker.sh:22]: `weaviate_version` can be removed

File: tools/dev/goreleaser_and_sign.sh [hunk 11]
```diff
@@ -3,18 +3,24 @@ set -eou pipefail
...
 # run from Weaviate Root!
 #
-# This script calls goreleaser and signs + notarize the binaries for macos
+# This script calls goreleaser and signs + notarize the binaries for macos. Thus needed to export few ENV variables
+# to be used in .goreleaser.yaml in the root of the repository.
 
-export GIT_HASH=$(git rev-parse --short HEAD)
+GIT_REVISION=$(git rev-parse --short HEAD)
+GIT_BRANCH=$(git rev-parse --abbrev-ref HEAD)
+BUILD_USER="$(whoami)@$(hostname)"
```
Comment (antas-marcin) [inline tools/dev/goreleaser_and_sign.sh:11]: let's make it a `ci` here maybe, I'm using this script to build the macOS binaries and attach them to release notes, I think `ci` here would be better (as we have it in `push_docker.sh` script
)

File: usecases/build/build.go (hunk 7 above)
Comment (antas-marcin) [reply usecases/build/build.go:25]: all is good :)

File: tools/dev/goreleaser_and_sign.sh (hunk 11 above)
Comment (kavirajk) [reply tools/dev/goreleaser_and_sign.sh:11]: Make sense 👍

File: usecases/config/config_handler.go (hunk 2 above)
Comment (antas-marcin) [reply usecases/config/config_handler.go:42]: we resolved this during our huddle :)

**Review by antas-marcin — APPROVED** [review]
